### PR TITLE
Refactor for Settings InputWidget and ButtonWidget.

### PR DESCRIPTION
**Pull request by robinr**
Following are present in the following PR:-
a) TextValidation Class contains the input text validation or any validation logic which
     is needed across the application under one place.
b) InputWidget and ButtonWidget which purely focus on the applicability in appropriate business logic.
c) AddCustomer and ErrorCatchWidget are examples of utilization of the items evolved  from the refactor (a) and (b).

d) Code cleanup.

### CODE REVIEW

File: components/playback/controls/player/AddCustomer.vue [hunk 1]
```diff
@@ -6,16 +6,54 @@
...
-        :placeholder="keyplaceholder_encode"
-        :spellcheck="spellchecking"
+        :title_label="'Enter Alphanumeric ID'"
+        :placeholder="'2VSckkBYr2An3dqHEyfRRE'"
+        :spellcheck="false"
         :invalid_text="error_text_uuid"
-        :input_width="entry_width"
-        :disabled="disallow_entry"
+        :input_width="400"
+        :disabled="false"
```
Comment (eli88fine) [inline components/playback/controls/player/AddCustomer.vue:14]: the default value for this prop is already `false`, so it's not necessary to specify that here

File: components/playback/controls/player/AddCustomer.vue [hunk 2]
```diff
@@ -6,16 +6,54 @@
...
     </div>
+
+    <div id="apikey" style="top: 140px; left: 50px; position: absolute">
+      <InputWidget
+        :title_label="'Enter API Key(Optional)'"
+        :placeholder="'ba86b8f0-6fdf-4944-87a0-8a491a19490e'"
+        :spellcheck="false"
+        :invalid_text="error_text_api"
+        :input_width="400"
+        :disabled="false"
```
Comment (eli88fine) [inline components/playback/controls/player/AddCustomer.vue:27]: the default value for this prop is already `false`, so it's not necessary to specify that here

File: components/playback/controls/player/AddCustomer.vue [hunk 3]
```diff
@@ -6,16 +6,54 @@
...
+      ></InputWidget>
+    </div>
+    <div id="nickname" style="top: 240px; left: 50px; position: absolute">
+      <InputWidget
+        :title_label="'Enter ID Nickname'"
+        :placeholder="'Curi Bio Main Account'"
+        :spellcheck="false"
+        :invalid_text="error_text_nickname"
+        :input_width="400"
+        :disabled="false"
```
Comment (eli88fine) [inline components/playback/controls/player/AddCustomer.vue:39]: the default value for this prop is already `false`, so it's not necessary to specify that here

File: components/playback/controls/player/AddCustomer.vue [hunk 4]
```diff
@@ -6,16 +6,54 @@
...
+        :dom_id_suffix="'apikey-id'"
+        @update:value="on_update_api($event)"
+      ></InputWidget>
+    </div>
+    <div id="nickname" style="top: 240px; left: 50px; position: absolute">
+      <InputWidget
+        :title_label="'Enter ID Nickname'"
+        :placeholder="'Curi Bio Main Account'"
+        :spellcheck="false"
```
Comment (eli88fine) [inline components/playback/controls/player/AddCustomer.vue:36]: The "Nickname" is typically something in plain English. Is there a reason you think spellcheck should be disabled for this field?

File: components/playback/controls/player/AddCustomer.vue [hunk 5]
```diff
@@ -25,54 +63,91 @@ import BootstrapVue from "bootstrap-vue";
...
+        user_ids: [],
+      };
+      this.$emit("save-id", add_customer);
+      this.$bvModal.hide("add-customer");
+    },
+    enable_save_button() {
+      if (this.error_text_uuid === "") {
+        if (this.error_text_api === "") {
+          if (this.error_text_nickname === "") {
+            this.enablelist_add_customer = [true, true];
```
Comment (eli88fine) [inline components/playback/controls/player/AddCustomer.vue:141]: you can add a `return` statement after setting `[true, true]`,  and then just have one line of code setting all other cases to `[true, false]`.  This will help deduplicate the code within this function

File: tests/unit/components/playback/controls/AddCustomer.spec.js [hunk 6]
```diff
@@ -120,9 +244,274 @@ describe("AddCustomer.enter_uuidbase57", () => {
       await Vue.nextTick();
       expect(spied_text_validator).toHaveBeenCalledWith(entry);
 
-      expect(target_error_message.text()).toStrictEqual(
-        spied_text_validator.mock.results[0].value
+      expect(target_error_message.text()).toStrictEqual(response_text);
```
Comment (eli88fine) [inline tests/unit/components/playback/controls/AddCustomer.spec.js:247]: why did you change this?  We had a specific discussion in Slack about trying to keep the tests in AddCustomer from being brittle by not testing the internal behavior of the TextValidation function
https://curibio.slack.com/archives/DQYMJLLAV/p1604414727006900

File: tests/unit/components/playback/controls/AddCustomer.spec.js [hunk 7]
```diff
@@ -120,9 +244,274 @@ describe("AddCustomer.enter_uuidbase57", () => {
...
+
+  test.each([
+    [
+      "0VSckkBYH2An3dqHEyfRRE",
+      "The entered ID has an invalid character 0,",
+      "06ad547f-fe02-477b-9473-f7977e4d5e17",
+      "",
+      "Experiment anemia -1",
+      "",
+      "color: rgb(63, 63, 63); width: 250px; left: 250px;",
```
Comment (eli88fine) [inline tests/unit/components/playback/controls/AddCustomer.spec.js:282]: the `width` and `left` attributes are not part of what you want to be testing here right?  Can you use `.contains` or some other assertion to just check that the color matches what you expect?

File: tests/unit/components/playback/controls/AddCustomer.spec.js [hunk 8]
```diff
@@ -120,9 +244,274 @@ describe("AddCustomer.enter_uuidbase57", () => {
...
+      "5FY8KwTsQaUJ2KzHJGetfE",
+      "",
+      "06ad547f-fe02-477b-9473-f7977e4d5e17",
+      "",
+      "Experiment anemia -1",
+      "",
+      "color: rgb(255, 255, 255); width: 250px; left: 250px;",
+    ],
+  ])(
+    "Given an UUID, API Key, Nickname for 'Add Customer' as input, When the input contains based on valid the critera or failure, Then display of Label 'Save ID' is visible or greyed",
```
Comment (eli88fine) [inline tests/unit/components/playback/controls/AddCustomer.spec.js:312]: When you use `test.each`, make sure to include some of the variables in the description (the `%s`).  This allows us to see from the Jest output if one of them fails which one it was

File: tests/unit/components/playback/controls/AddCustomer.spec.js [hunk 9]
```diff
@@ -120,9 +244,274 @@ describe("AddCustomer.enter_uuidbase57", () => {
...
+      const selector_id_suffix_alphanumeric_id = "alphanumeric-id";
+      const selector_id_suffix_apikey_id = "apikey-id";
+      const selector_id_suffix_nickname_id = "nickname-id";
+
+      const spied_text_validator_uuid = jest.spyOn(
+        TextValidation.prototype,
+        "validate_uuidBase_fiftyseven_encode"
+      );
+
+      const spied_text_validator_api = jest.spyOn(
```
Comment (eli88fine) [inline tests/unit/components/playback/controls/AddCustomer.spec.js:331]: the use of these validators is already being tested elsewhere.  To try and make our tests less brittle (i.e. not have to refactor them unnecessarily), please try and only include things within a test that are necessary to test the specific "unit" being tested.

In this test, you are not checking the validation logic of individual fields----you are checking whether the Save button is enabled/disabled when certain fields are valid/invalid

File: tests/unit/components/playback/controls/AddCustomer.spec.js [hunk 10]
```diff
@@ -120,9 +244,274 @@ describe("AddCustomer.enter_uuidbase57", () => {
...
+      "06ad547f-fe02-477b-9473-f7977e4d5e17",
+      "",
+      "Experiment anemia -1",
+      "",
+      "color: rgb(255, 255, 255); width: 250px; left: 250px;",
+      1,
+    ],
+  ])(
+    "Given an UUID, API Key, Nickname for 'Add Customer' as input, When the input contains based on valid the critera or failure, Then display of Label 'Save ID' is visible, click on Cancel, emitted event (value 0) and click on Save emitted event (value 1)",
+    async (
```
Comment (eli88fine) [inline tests/unit/components/playback/controls/AddCustomer.spec.js:422]: all the same comments as the previous test apply to this one as well

File: tests/unit/components/playback/controls/AddCustomer.spec.js [hunk 11]
```diff
@@ -120,9 +244,274 @@ describe("AddCustomer.enter_uuidbase57", () => {
...
+
+      await save_btn.trigger("click");
+      await Vue.nextTick();
+      expect(spied_button_function).toHaveBeenCalledWith(label);
+
+      const save_id_events = wrapper.emitted("save-id");
+      expect(save_id_events).toHaveLength(1);
+      expect(save_id_events[0]).toStrictEqual([
+        {
+          api_key: "06ad547f-fe02-477b-9473-f7977e4d5e17",
```
Comment (eli88fine) [inline tests/unit/components/playback/controls/AddCustomer.spec.js:508]: have these values use the `test.each` variables you created at the top of the test

File: tests/unit/components/playback/controls/InputWidget.spec.js [hunk 12]
```diff
@@ -225,4 +225,27 @@ describe("InputWidget.vue", () => {
       "width: 390px; top: 48px;"
     );
   });
+  test("When the component is mounted, Then then the props value 'display_text_message' is set to 'false' the invalid_text is not rendered, on resetting 'display_text_message' to 'true' then invalid_text is rendered", async () => {
```
Comment (eli88fine) [inline tests/unit/components/playback/controls/InputWidget.spec.js:228]: an initial props configuration should not be in the `Then` section.  The configuration of the props are part of wherever you think the `mounted` should be (either in `Given` or `When`, depending on the test).

So something like `When the component is mounted with the display_text_message prop set to false, Then the invalid_text is not rendered`

Also, you appear to be testing two separate things here (your "Then" clause has become quite long).

I would recommend creating a separate second test: `Given the component was mounted with the display_text_message prop set to false, When the display_text_message prop is updated to True, Then the invalid_text is rendered`

**Review by eli88fine — CHANGES_REQUESTED** [review]
Great work overall. Some minor comments to address before merge

File: components/playback/controls/player/AddCustomer.vue (hunk 1 above)
Comment (robinr) [reply components/playback/controls/player/AddCustomer.vue]: Agreed.

File: components/playback/controls/player/AddCustomer.vue (hunk 2 above)
Comment (robinr) [reply components/playback/controls/player/AddCustomer.vue]: Agreed

File: components/playback/controls/player/AddCustomer.vue (hunk 3 above)
Comment (robinr) [reply components/playback/controls/player/AddCustomer.vue]: Agreed

File: components/playback/controls/player/AddCustomer.vue (hunk 4 above)
Comment (robinr) [reply components/playback/controls/player/AddCustomer.vue:34]: Spellcheck is to verify spelling and grammar checked, if set not set 'false' by default it would be 'true' and there is no UI/UX to show a squiggle text under nickname, which will get applied so its essential to have spellcheck="false" for nickname.

File: components/playback/controls/player/AddCustomer.vue (hunk 5 above)
Comment (robinr) [reply components/playback/controls/player/AddCustomer.vue:138]: Completed

File: tests/unit/components/playback/controls/AddCustomer.spec.js (hunk 6 above)
Comment (robinr) [reply tests/unit/components/playback/controls/AddCustomer.spec.js]: Completed.

File: tests/unit/components/playback/controls/AddCustomer.spec.js (hunk 7 above)
Comment (robinr) [reply tests/unit/components/playback/controls/AddCustomer.spec.js:244]: https://vue-test-utils.vuejs.org/api/wrapper/#contains 

As per your earlier review never to use functions which have "Deprecation warning".
So as per earlier comment I haven't used .contains.

File: tests/unit/components/playback/controls/AddCustomer.spec.js (hunk 8 above)
Comment (robinr) [reply tests/unit/components/playback/controls/AddCustomer.spec.js]: Completed.

File: tests/unit/components/playback/controls/AddCustomer.spec.js (hunk 9 above)
Comment (robinr) [reply tests/unit/components/playback/controls/AddCustomer.spec.js]: Agreed and removed.

File: tests/unit/components/playback/controls/AddCustomer.spec.js (hunk 10 above)
Comment (robinr) [reply tests/unit/components/playback/controls/AddCustomer.spec.js:369]: Agreed

File: tests/unit/components/playback/controls/AddCustomer.spec.js (hunk 11 above)
Comment (robinr) [reply tests/unit/components/playback/controls/AddCustomer.spec.js]: Completed.

File: tests/unit/components/playback/controls/InputWidget.spec.js (hunk 12 above)
Comment (robinr) [reply tests/unit/components/playback/controls/InputWidget.spec.js]: Completed.

**Comment by robinr** [comment]
All the Review comments implemented and the feature of cut_paste_disable if set to true to disable CTRL+V done.

File: components/playback/controls/player/AddCustomer.vue (hunk 4 above)
Comment (eli88fine) [reply components/playback/controls/player/AddCustomer.vue:34]: per the conversation with Ari on Slack today, please make a note to update `spellcheck` to true (or just remove it since the default is true) as part of your next pull request

File: tests/unit/components/playback/controls/AddCustomer.spec.js [hunk 13]
```diff
@@ -123,6 +206,244 @@ describe("AddCustomer.enter_uuidbase57", () => {
       expect(target_error_message.text()).toStrictEqual(
         spied_text_validator.mock.results[0].value
       );
+      // expect(target_error_message.text()).toStrictEqual(response_text);
```
Comment (eli88fine) [inline tests/unit/components/playback/controls/AddCustomer.spec.js:209]: remove old commented code

File: tests/unit/components/playback/controls/AddCustomer.spec.js (hunk 7 above)
Comment (eli88fine) [reply tests/unit/components/playback/controls/AddCustomer.spec.js:244]: Sorry, I should have clarified.  I meant the Jest assertion, not the Vue Test Utils wrapper.  https://jestjs.io/docs/en/expect#tocontainitem

`expect(cancel_btn.attributes().style).toContain("color: rgb(63, 63, 63);")`

File: components/playback/controls/player/AddCustomer.vue (hunk 5 above)
Comment (eli88fine) [reply components/playback/controls/player/AddCustomer.vue:138]: the following is what I meant to reduce the number of duplicated statements

`if (this.error_text_uuid === "") {`

`    if (this.error_text_api === "") {`

`          if (this.error_text_nickname === "") {`

`            this.enablelist_add_customer = [true, true];`

`           return`

`          }`

`     }`

`}`

`this.enablelist_add_customer = [true, false];`

File: components/playback/controls/player/InputWidget.vue [hunk 14]
```diff
@@ -92,6 +94,8 @@ export default {
     input_width: { type: Number, default: 0 }, // textbox_width (int)  [pixels]
     disabled: { type: Boolean, default: false }, // disabled (optional bool=False) (not able to type into input)
     dom_id_suffix: { type: String, default: "" }, // TODO (Eli 11/3/20): consider defaulting this to a random UUID if no value supplied
+    display_text_message: { type: Boolean, default: true }, // display_text_message (boolean) if set to false would not render invalid_text
+    cut_paste_disable: { type: Boolean, default: false }, // cut_paste_disable (boolean) if set to true would prevent cut and paste of text into input
```
Comment (eli88fine) [inline components/playback/controls/player/InputWidget.vue:98]: this only disables pasting right? it doesn't disable the "cut" inside the text input field, correct?
If that's true, then I think it should only be named `disable_paste`.  There may be other situations in the future where for some reason we want to disable cut inside the text field, but that is not the current situation

**Review by eli88fine — APPROVED** [review]
Approved. Please make notes to address the minor comments as part of your next Pull Request

File: tests/unit/components/playback/controls/AddCustomer.spec.js (hunk 13 above)
Comment (robinr) [reply tests/unit/components/playback/controls/AddCustomer.spec.js:209]: Done.

File: components/playback/controls/player/InputWidget.vue (hunk 14 above)
Comment (robinr) [reply components/playback/controls/player/InputWidget.vue:98]: Modified the api and the prop name to disable_paste

File: components/playback/controls/player/AddCustomer.vue [hunk 15]
```diff
@@ -8,7 +8,6 @@
       <InputWidget
         :title_label="'Enter Alphanumeric ID'"
```
Comment (eli88fine) [inline components/playback/controls/player/AddCustomer.vue:9]: The alphanumeric ID and API key should still have `spellcheck='false'`.  Those are not fields expected to contain English words.  Just the Nickname should use spellchecking

**Review by eli88fine — APPROVED** [review]
Please just make this minor correction in the branch you work on for your next task.

File: components/playback/controls/player/AddCustomer.vue (hunk 15 above)
Comment (robinr) [reply components/playback/controls/player/AddCustomer.vue:9]: ok.
